### PR TITLE
[DUOS-1764, DUOS-1765][risk=no] Create a single DAR for N Datasets on submission

### PR DESCRIPTION
**Pull request by rushtong**
## Addresses
https://broadworkbench.atlassian.net/browse/DUOS-1764
https://broadworkbench.atlassian.net/browse/DUOS-1765

## Changes
* Update to the DAR submission process so we only create a single DAR instead of N DARs per dataset requested.
* Updates to resource class and tests to cover this change.

Note that we should immediately follow up on https://broadworkbench.atlassian.net/browse/DUOS-1961 as that story is blocked by this one and we will no longer need to archive a group of DARs (DARCollection) and recreate a new draft from them.

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/java/org/broadinstitute/consent/http/db/DataAccessRequestDAO.java [hunk 1]
```diff
@@ -318,20 +317,4 @@ void insertDataAccessRequest(
...
-  @SqlQuery("SELECT distinct reference_id, dataset_id FROM dar_dataset WHERE reference_id IN (<referenceIds>)")
-  List<DarDataset> findAllDARDatasets(@BindList("referenceIds") List<String> referenceIds);
-
-  @SqlQuery(
-      " SELECT distinct d.reference_id "
-          + " FROM dar_dataset d "
-          + " INNER JOIN data_access_request dar ON dar.reference_id = d.reference_id AND dar.collection_id = :collectionId "
-          + " WHERE d.dataset_id IN <datasetIds> ")
-  List<String> findReferenceIdsForDatasetIdsWithCollectionId(
-      @BindList("datasetIds") List<Integer> datasetIds, @Bind("collectionId") Integer collectionId);
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/db/DataAccessRequestDAO.java:335]: Unused

File: src/main/java/org/broadinstitute/consent/http/db/ElectionDAO.java [hunk 2]
```diff
@@ -268,6 +270,19 @@ List<Election> findLastElectionsByReferenceIds(
     @UseRowMapper(SimpleElectionMapper.class)
     Election findLastElectionByReferenceIdAndType(@Bind("referenceId") String referenceId, @Bind("type") String type);
 
+    @SqlQuery(
+        " SELECT e.* " +
+        " FROM election e " +
+        " INNER JOIN (SELECT referenceid, datasetid, MAX(createdate) max_date FROM election WHERE LOWER(electiontype) = lower(:type) AND datasetid = :datasetId GROUP BY referenceid, datasetid) election_view " +
```
Comment (JVThomas) [inline src/main/java/org/broadinstitute/consent/http/db/ElectionDAO.java:276]: Why use MAX(createdate) over something like MAX(electionid)?

Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/db/ElectionDAO.java:276]: Mostly because we use it a lot in other queries. I believe both approaches will work equally well.

**Review by shaemarks — COMMENTED** [review]
Looks good overall! When run locally with the ui, I submitted a multi-dataset dar, and the resulting collection contained the single dar with all of the datasets.

However, I think some changes will need to be made to the bucketing logic on the frontend to support this change. The collection I created contained 4 datause buckets, each containing all of the elections since they are processed by dar. The votes are then processed by elections, so voting shouldn't be accurate with these changes.

processDataUseBuckets

**Comment by rushtong** [comment]
> However, I think some changes will need to be made to the bucketing logic on the frontend to support this change. The collection I created contained 4 datause buckets, each containing all of the elections since they are processed by dar. The votes are then processed by elections, so voting shouldn't be accurate with these changes.

I did actually test this case, but the bucketing and votes looked correct to me when I did so. That could have been an edge case where it worked, and there are others where it doesn't ... I'll try to dig more into the front end and figure something out.

File: src/main/java/org/broadinstitute/consent/http/service/dao/DarCollectionServiceDAO.java [hunk 3]
```diff
@@ -79,59 +79,62 @@ public List<String> createElectionsForDarCollection(User user, DarCollection col
           //    4. Member votes for rp election
           //        4a. Chair Vote for chair
           collection.getDars().values().forEach(dar -> {
-                // If there is an existing open election for this DAR, we can ignore it
-                Election lastDataAccessElection = electionDAO.findLastElectionByReferenceIdAndType(dar.getReferenceId(), ElectionType.DATA_ACCESS.getValue());
-                boolean ignore = Objects.nonNull(lastDataAccessElection) && lastDataAccessElection.getStatus().equals(ElectionStatus.OPEN.getValue());
+                dar.getDatasetIds().forEach(datasetId -> {
+                    // If there is an existing open election for this DAR+Dataset, we can ignore it
+                    Election lastDataAccessElection = electionDAO.findLastElectionByReferenceIdDatasetIdAndType(dar.getReferenceId(), datasetId, ElectionType.DATA_ACCESS.getValue());
+                    boolean ignore = Objects.nonNull(lastDataAccessElection) && lastDataAccessElection.getStatus().equals(ElectionStatus.OPEN.getValue());
```
Comment (JVThomas) [inline src/main/java/org/broadinstitute/consent/http/service/dao/DarCollectionServiceDAO.java]: Nit: Just to be safe, can you update the old `equals` usage with `equalsIgnoreCase`?

**Review by JVThomas — APPROVED** [review]
Minor nitpick, but the code looks good. That said, Shae is right about the bucketing logic on the front-end. It needs to be updated to account for the "one dar to many elections" relationship. An update there needs to happen before this PR can go to staging/production.

**Review by shaemarks — APPROVED** [review]
